File: RadarScope/radar_math.py
"""
Radar math utilities for coordinate conversion and bearing/distance calculations.
"""

import math

def haversine_distance(lat1, lon1, lat2, lon2):
    """
    Calculate the great-circle distance between two points on Earth.
    Returns distance in nautical miles.
    """
    R = 3440.065  # Earth radius in nautical miles
    dlat = math.radians(lat2 - lat1)
    dlon = math.radians(lon2 - lon1)
    a = math.sin(dlat/2)**2 + math.cos(math.radians(lat1)) * math.cos(math.radians(lat2)) * math.sin(dlon/2)**2
    c = 2 * math.atan2(math.sqrt(a), math.sqrt(1-a))
    return R * c

def calculate_bearing(lat1, lon1, lat2, lon2):
    """
    Calculate the bearing from point 1 to point 2.
    Returns bearing in degrees from north (0-360).
    """
    lat1 = math.radians(lat1)
    lon1 = math.radians(lon1)
    lat2 = math.radians(lat2)
    lon2 = math.radians(lon2)

    dlon = lon2 - lon1
    y = math.sin(dlon) * math.cos(lat2)
    x = math.cos(lat1) * math.sin(lat2) - math.sin(lat1) * math.cos(lat2) * math.cos(dlon)
    bearing = math.atan2(y, x)
    bearing = math.degrees(bearing)
    bearing = (bearing + 360) % 360
    return bearing

def radar_coordinates(distance_nm, bearing_deg, radar_radius_nm):
    """
    Convert polar coordinates (distance, bearing) to radar screen coordinates.
    Assumes radar center is at (0,0) and y-axis points north, x-axis points east.
    Returns (x, y) in pixels, assuming radar_radius_nm maps to radar_radius_pixels.
    """
    # Convert bearing from degrees to radians (0° is north, 90° is east)
    theta = math.radians(bearing_deg)
    # Convert distance to radar radius ratio (0 to 1)
    ratio = distance_nm / radar_radius_nm if radar_radius_nm > 0 else 0
    # Clamp ratio to 1.0 (maximum display range)
    ratio = min(ratio, 1.0)
    # Calculate coordinates (y is north, x is east)
    x = ratio * math.sin(theta)  # easting
    y = ratio * math.cos(theta)  # northing
    return x, y

def interpolate_position(old_pos, new_pos, factor):
    """
    Linearly interpolate between two positions.
    factor: 0.0 (old_pos) to 1.0 (new_pos)
    """
    if old_pos is None:
        return new_pos
    if new_pos is None:
        return old_pos
    lat = old_pos[0] + (new_pos[0] - old_pos[0]) * factor
    lon = old_pos[1] + (new_pos[1] - old_pos[1]) * factor
    return (lat, lon)

File: RadarScope/static/js/radar_math.js
// Radar math utilities for coordinate conversion and bearing/distance calculations.

/**
 * Convert degrees to radians
 * @param {number} degrees - Angle in degrees
 * @returns {number} Angle in radians
 */
function toRadians(degrees) {
    return degrees * Math.PI / 180;
}

/**
 * Convert radians to degrees
 * @param {number} radians - Angle in radians
 * @returns {number} Angle in degrees
 */
function toDegrees(radians) {
    return radians * 180 / Math.PI;
}

/**
 * Calculate the great-circle distance between two points on Earth.
 * @param {number} lat1 - Latitude of point 1 in degrees
 * @param {number} lon1 - Longitude of point 1 in degrees
 * @param {number} lat2 - Latitude of point 2 in degrees
 * @param {number} lon2 - Longitude of point 2 in degrees
 * @returns {number} Distance in nautical miles
 */
function haversineDistance(lat1, lon1, lat2, lon2) {
    const R = 3440.065; // Earth radius in nautical miles
    const dLat = toRadians(lat2 - lat1);
    const dLon = toRadians(lon2 - lon1);
    const lat1Rad = toRadians(lat1);
    const lat2Rad = toRadians(lat2);
    const a = Math.sin(dLat/2) * Math.sin(dLat/2) +
              Math.cos(lat1Rad) * Math.cos(lat2Rad) *
              Math.sin(dLon/2) * Math.sin(dLon/2);
    const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1-a));
    return R * c;
}

/**
 * Calculate the bearing from point 1 to point 2.
 * @param {number} lat1 - Latitude of point 1 in degrees
 * @param {number} lon1 - Longitude of point 1 in degrees
 * @param {number} lat2 - Latitude of point 2 in degrees
 * @param {number} lon2 - Longitude of point 2 in degrees
 * @returns {number} Bearing in degrees from north (0-360)
 */
function calculateBearing(lat1, lon1, lat2, lon2) {
    const lat1Rad = toRadians(lat1);
    const lat2Rad = toRadians(lat2);
    const dLon = toRadians(lon2 - lon1);
    const y = Math.sin(dLon) * Math.cos(lat2Rad);
    const x = Math.cos(lat1Rad) * Math.sin(lat2Rad) -
              Math.sin(lat1Rad) * Math.cos(lat2Rad) * Math.cos(dLon);
    let bearing = Math.atan2(y, x);
    bearing = toDegrees(bearing);
    bearing = (bearing + 360) % 360;
    return bearing;
}

/**
 * Convert polar coordinates (distance, bearing) to radar screen coordinates.
 * Assumes radar center is at (0,0) and y-axis points north, x-axis points east.
 * @param {number} distanceNm - Distance in nautical miles from center
 * @param {number} bearingDeg - Bearing in degrees from north (0-360)
 * @param {number} radarRadiusNm - Radar radius in nautical miles (max display range)
 * @param {number} radarRadiusPx - Radar radius in pixels (screen radius)
 * @returns {{x: number, y: number}} Radar coordinates (x, y) in pixels, where (0,0) is center
 */
function radarCoordinates(distanceNm, bearingDeg, radarRadiusNm, radarRadiusPx) {
    // Convert bearing from degrees to radians (0° is north, 90° is east)
    const theta = toRadians(bearingDeg);
    // Convert distance to radar radius ratio (0 to 1)
    let ratio = distanceNm / radarRadiusNm;
    // Clamp ratio to 1.0 (maximum display range)
    if (ratio > 1) ratio = 1;
    // Calculate coordinates (y is north, x is east)
    const x = ratio * Math.sin(theta) * radarRadiusPx; // easting
    const y = ratio * Math.cos(theta) * radarRadiusPx; // northing
    return { x, y };
}

/**
 * Linearly interpolate between two positions.
 * @param {{latitude: number, longitude: number}} oldPos - Old position
 * @param {{latitude: number, longitude: number}} newPos - New position
 * @param {number} factor - Interpolation factor (0.0 = oldPos, 1.0 = newPos)
 * @returns {{latitude: number, longitude: number}} Interpolated position
 */
function interpolatePosition(oldPos, newPos, factor) {
    if (!oldPos) return newPos;
    if (!newPos) return oldPos;
    const lat = oldPos.latitude + (newPos.latitude - oldPos.latitude) * factor;
    const lon = oldPos.longitude + (newPos.longitude - oldPos.longitude) * factor;
    return { latitude: lat, longitude: lon };
}

// Export for use in other modules (if using modules)
// In a browser environment, we'll attach to window
window.radarMath = {
    toRadians,
    toDegrees,
    haversineDistance,
    calculateBearing,
    radarCoordinates,
    interpolatePosition
};